text/0000-natvis.md

```diff
@@ -0,0 +1,495 @@
+would be redundant. Currently, running the Debug trait in the debugger directly
+is not possible and so a manual definition would be required to have a debugger view.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+This design provides a simple mechanism for cargo to collect the list of
+.natvis files specified for a given crate and embed them in the resulting
+pdb. It does not need any manual intervention by a Rust developer who is
+consuming such a crate to get the debugging experience to work when it is
+viewed under a debugger that supports the Natvis Framework.
+
+This design does not break any existing usage of cargo or rustc. This new feature
+would be strictly opt-in. The Natvis syntax may not be familiar to many Rust
+developers which may lead to a period of learning the syntax. Since this feature
+would be optional, a consumer of a crate that has natvis definitions for types
+would not need to go through this learning curve. 
+
+Not doing this would keep the existing debugging experience for external Rust crates.
+Most Rust types, outside of the standard library, do not have any debugger views
+defined for them by default which makes them difficult to interpret when viewed
+under a debugger.
+
+# Prior art
+[prior-art]: #prior-art
+
+Many debuggers and languages already address this problem. Some do so in a way
+that is more flexible than others.
+
+Briefly, we cover some of the known systems for debugger views:
+
+* Windows Debugger (WinDbg)
+* Visual Studio Debugger (VS Debugger)
+* GDB/LLDB
+
+## Windows Debugger (WinDbg)
+
+Natvis is a framework that customizes how native types appear when viewed under
+a debugger. The Visual Studio Natvis framework is supported out of the box on
+WinDbg. The debugger has the ability to load `.natvis` files via the `.nvload`
+command and directly apply them to types within loaded modules. WinDbg is also
+able to load `.natvis` files that have been embedded in the PDB for a binary and
+serve up the resulting views after applying those visualizations as well. This
+allows for a very smooth debugging experience which would not depend on any manual
+loading of Natvis files.
+
+## Visual Studio Debugger (VS Debugger)
+
+The Visual Studio Debugger also supports Natvis. Similar to WinDbg, the VS Debugger
+is also able to apply Natvis on the fly by loading user-specified `.natvis` files.
+As with WinDbg, it also supports loading `.natvis` files that were embedded in the
+PDB for a binary and automatically applying the Natvis visualizations to types from
+that binary.
+
+When using Visual Studio to build a C++ project, a developer can add a Natvis file
+via the `.vcxproj` file. To add a Natvis file to a project the following can be
+added to the `.vcxproj` file:
+
+```text
+<ItemGroup>
+  <Natvis Include="Foo.natvis" />
+</ItemGroup>
+```
+
+## GDB/LLDB
+
+GDB and LLDB also support debugger views but in a different way than WinDbg and the
```

Comment:
Just to add additional context here since I saw other people asking for more information about how other debuggers work, `gdb` supports a similar concept where a special section is generated that contains the name of debugger scripts to load and execute when a binary is being debugged (see [here](https://github.com/rust-lang/rust/blob/89b9f7b284aacc5f8613438b80e4dd7bdd10549e/compiler/rustc_codegen_llvm/src/debuginfo/gdb.rs#L29-L59) where we generate this section). Gdb also seems to support embedding the scripts directly into this section as well (https://sourceware.org/gdb/onlinedocs/gdb/dotdebug_005fgdb_005fscripts-section.html).